src/JuMPContainer.jl

```diff
@@ -221,139 +221,20 @@ else
 end
 Base.length(it::ValueIterator) = length(it.x)
 
-mutable struct KeyIterator{JA<:JuMPArray}
-    x::JA
-    dim::Int
-    next_k_cache::Array{Any,1}
-    function KeyIterator{JA}(d) where JA
-        n = ndims(d.innerArray)
-        new{JA}(d, n, Array{Any}(undef, VERSION < v"0.7-" ? n+1 : n))
+struct KeyIterator{T}
```

Comment:
If we decide to change the type of the iterator, we might as well return `Iterators.product(d.indexsets...)` directly in `keys` and not have `KeyIterator`.

Reply:
The only drawback thereof is that Iterators.ProductIterator sets the HasShape trait and defines its size as a tuple of dimensions. For instance, `size(product(1:2, 1:3)) = (2, 3)`, s.t. `collect(product(1:2, 1:3))` becomes
```
julia> collect(Iterators.product(1:2, 3:5))
2×3 Array{Tuple{Int64,Int64},2}:
 (1, 3)  (1, 4)  (1, 5)
 (2, 3)  (2, 4)  (2, 5)
```
contrarily to what the current and previous `KeyIterator` returned
```
julia> collect(keys(JuMPArray(rand(2,3), (1:2, 1:3))))
6-element Array{Any,1}:
 (1, 1)
 (2, 1)
 (1, 2)
 (2, 2)
 (1, 3)
 (2, 3)
```

For its main purpose, i.e. the iteration over all index pairs, that would not have any impact, of course.